nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/CoreService.java

```diff
@@ -125,4 +141,97 @@ public VersionRemovalPolicy getVersionRemovalPolicy() {
+     * @return the number of orphan versions deleted
+     * @since 9.1
+     */
+    public long cleanupOrphanVersions(long commitSize) {
+        RepositoryService repositoryService = Framework.getService(RepositoryService.class);
+        if (repositoryService == null) {
+            // not initialized
+            return 0;
+        }
+        List<String> repositoryNames = repositoryService.getRepositoryNames();
+        AtomicLong count = new AtomicLong();
+        for (String repositoryName : repositoryNames) {
+            TransactionHelper.runInTransaction(() -> {
+                CoreInstance.doPrivileged(repositoryName, (CoreSession session) -> {
+                    count.addAndGet(doCleanupOrphanVersions(session, commitSize));
+                });
+            });
+        }
+        return count.get();
+    }
+
+    protected long doCleanupOrphanVersions(CoreSession session, long commitSize) {
+        // compute map of version series -> list of version ids in it
+        Map<String, List<String>> versionSeriesToVersionIds = new HashMap<>();
+        String findVersions = "SELECT " + NXQL.ECM_UUID + ", " + NXQL.ECM_VERSION_VERSIONABLEID
+                + " FROM Document WHERE " + NXQL.ECM_ISVERSION + " = 1";
+        try (IterableQueryResult res = session.queryAndFetch(findVersions, NXQL.NXQL)) {
+            for (Map<String, Serializable> map : res) {
+                String versionSeriesId = (String) map.get(NXQL.ECM_VERSION_VERSIONABLEID);
+                String versionId = (String) map.get(NXQL.ECM_UUID);
+                versionSeriesToVersionIds.computeIfAbsent(versionSeriesId, k -> new ArrayList<>(4)).add(versionId);
+            }
+        }
+        Set<String> seriesIds = new HashSet<>();
+        // find the live doc ids
+        String findLive = "SELECT " + NXQL.ECM_UUID + " FROM Document WHERE " + NXQL.ECM_ISPROXY + " = 0 AND "
+                + NXQL.ECM_ISVERSION + " = 0";
+        try (IterableQueryResult res = session.queryAndFetch(findLive, NXQL.NXQL)) {
+            for (Map<String, Serializable> map : res) {
+                String id = (String) map.get(NXQL.ECM_UUID);
+                seriesIds.add(id);
+            }
+        }
+        // find the version series for proxies
+        String findProxies = "SELECT " + NXQL.ECM_PROXY_VERSIONABLEID + " FROM Document WHERE " + NXQL.ECM_ISPROXY
+                + " = 1";
+        try (IterableQueryResult res = session.queryAndFetch(findProxies, NXQL.NXQL)) {
+            for (Map<String, Serializable> map : res) {
+                String versionSeriesId = (String) map.get(NXQL.ECM_PROXY_VERSIONABLEID);
+                seriesIds.add(versionSeriesId);
+            }
+        }
+        // all version for series ids not found from live docs or proxies can be removed
+        versionSeriesToVersionIds.keySet().removeAll(seriesIds);
+        Set<String> ids = new HashSet<>();
+        for (Entry<String, List<String>> es : versionSeriesToVersionIds.entrySet()) {
```

Comment:
I think `e` or `entry` is better than `es`, because when using iteration, only one element / entry of the entry set is being used.